### PR TITLE
Hopwood Optimized Fixed-Base Scalar Multiplication

### PR DESCRIPTION
**Pull request by 95DDB**
Import Hopwood's optimized vbSM gadget as per https://github.com/zcash/zcash/issues/3924 .
Currently performed with incomplete arithmetic

### CODE REVIEW

File: r1cs/gadgets/std/src/groups/mod.rs [hunk 1]
```diff
@@ -231,6 +231,70 @@ pub(crate) fn check_mul_bits_inputs<
     Ok(())
 }
 
+/// Pre-checks for vbSM with incomplete arithmetic using Hopwood algorithm (https://github.com/zcash/zcash/issues/3924) :
+/// - 'self' must be non-trivial and in the prime order subgroup 
+/// - 'bits', in little endian, must be of length <= than the scalar field modulus. 
+///           and must not be equal to {0, p-2, p-1, p, p+1}.
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/std/src/groups/mod.rs]: outdated comment, I will change it.

File: r1cs/gadgets/std/src/groups/mod.rs [hunk 2]
```diff
@@ -231,6 +231,70 @@ pub(crate) fn check_mul_bits_inputs<
...
+        return Err(SynthesisError::Other("Invalid base point".to_owned()));
+    }
+
+    // Read FE from bits
+    let bits_val = G::ScalarField::read_bits(bits.clone())
+        .map_err(|e| SynthesisError::Other(e.to_string()))?;
+
+    let one = G::ScalarField::one();
+    let two = one.double();
+    let two_to_n = two.pow(&[bits_len as u64]);
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/std/src/groups/mod.rs]: n is the length of the *padded* bit sequence.

File: r1cs/gadgets/std/src/groups/mod.rs [hunk 3]
```diff
@@ -231,6 +231,70 @@ pub(crate) fn check_mul_bits_inputs<
...
+    if bits_val.double() == three_times_two_to_n_minus_one {
+        return Err(SynthesisError::Other("2 *  [b_n-1, ..., b_1, b_0] != 3*(2^n - 1)".to_owned()))
+    }
+
+    // 2 *  [b_n-1, ..., b_1, b_0] != [b_n-1, b_n-2] * (2^n) - 3
+    let msb_val = G::ScalarField::read_bits((&bits[..2]).to_vec()).unwrap();
+    let rhs = (msb_val * &two_to_n) - &three;
+    if bits_val.double() == rhs {
+        return Err(SynthesisError::Other("2 *  [b_n-1, ..., b_1, b_0] != [b_n-1, b_n-2] * (2^n) - 3".to_owned()))
+    }
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/std/src/groups/mod.rs]: As DDT hopefully informed you, I missed to communicate you a fourth condition:
```
[b_{n-1},...,b_0] = 0 mod p
```

File: r1cs/gadgets/std/src/groups/curves/short_weierstrass/short_weierstrass_jacobian.rs [hunk 4]
```diff
@@ -774,16 +790,29 @@ for AffineGadget<P, ConstraintF, F>
...
+            }
+
+            // Update values for next chunk
             t = t.double().double();
-            to_sub += &sigma;
+            to_sub += &table[0];
         }
-        acc = acc.sub_constant(cs.ns(|| "acc - sigma*n_div_2"), &to_sub)?;
-        acc = acc.sub(cs.ns(|| "subtract shift"), &shift)?;
+        acc = acc.sub_constant(cs.ns(|| "acc - to_sub"), &to_sub)?;
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/std/src/groups/curves/short_weierstrass/short_weierstrass_jacobian.rs:820]: I suppose sub_constant is safe, i.e. it enforces different x-coordinates?

**Review by UlrichHaboeck75 — CHANGES_REQUESTED** [review]
The checks for exceptions of the mul_bits_fixed_base needs some changes, as described in my inline comments.

File: r1cs/gadgets/std/src/groups/curves/short_weierstrass/short_weierstrass_jacobian.rs (hunk 4 above)
Comment (95DDB) [reply r1cs/gadgets/std/src/groups/curves/short_weierstrass/short_weierstrass_jacobian.rs:820]: Yes. It calls the add_constant which is only in the safe variant

**Review by UlrichHaboeck75 — APPROVED** [review]

**Pull request merged**